[rush] Using "includeDevDependencies": true for a deployment scenario fails for not found module from devDependencies of a dependency out of repository scope

## Summary

I want to run the `deploy` command during Docker build to prepare an image for running tests. The testing framework is defined in my project's `package.json` file in `devDependencies` as expected. However, the `deploy` command fails with `Cannot find module 'dev_dependency_module_of_a_dependency' from 'node_modules_path_to_the_dependency'`.

## Repro steps

1. Take any project having a dependency or devDependency having its own "devDependencies". I expect it is almost every modern package.
2. Prepare a deployment scenario having `includeDevDependencies` turned on.
3. Run `rush deploy` for that scenario
4. See the error message in console right after `Analyzing project: XYZ`

  **Expected result:** Not failing for missing devDependencies for any package out of current repo/not managed by rush.

  **Actual result:** Error message `Cannot find module 'dev_dependency_module_of_a_dependency' from 'node_modules_path_to_the_dependency'`

## Details

I believe the root cause is [this method](https://github.com/microsoft/rushstack/blob/250d7b40d9d811fc20a9c25a5f87abc60e70e9f7/apps/rush-lib/src/logic/deploy/DeployManager.ts#L142-L242) is recursive not checking the `packageJson` being processed belongs to a package managed by rush for devDependencies [here](https://github.com/microsoft/rushstack/blob/250d7b40d9d811fc20a9c25a5f87abc60e70e9f7/apps/rush-lib/src/logic/deploy/DeployManager.ts#L173-L177).

I guess it might work better for other package managers (NPM, Yarn) because some of these dependencies might be included in "dependencies" of some other package installed so it is found because of its "phantom friendly" node_modules structure. However, I believe that case reduces the probability of this case and does not mitigate it fully.

I wonder what devDependencies should be included anyway:
1. devDeps of the main project only?
2. devDeps of the main project and all `additionalProjectsToInclude` as well?
3. devDeps of all projects managed by rush?
4. devDeps of all projects in the respository? (I guess this is not an option because rush cannot work with other projects than managed.)
5. devDeps of all packages installed? (This is the current behavior.)

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| `@microsoft/rush` globally installed version? | 5.40.5 |
| `rushVersion` from rush.json? | 5.42.1 |
| `useWorkspaces` from rush.json? | false |
| Package manager? | pnpm |
| Package manager version? | 5.18.6 |
| Operating system? | Mac |
| Would you consider contributing a PR? | Yes |
| Node.js version (`node -v`)? | 12.20.1 |

--- a/apps/rush-lib/src/logic/deploy/DeployManager.ts
+++ b/apps/rush-lib/src/logic/deploy/DeployManager.ts
@@ -170,7 +170,7 @@ export class DeployManager {
     for (const name of Object.keys(packageJson.dependencies || {})) {
       dependencyNamesToProcess.add(name);
     }
-    if (deployState.scenarioConfiguration.json.includeDevDependencies) {
+    if (deployState.scenarioConfiguration.json.includeDevDependencies && sourceFolderInfo?.isRushProject) {
       for (const name of Object.keys(packageJson.devDependencies || {})) {
         dependencyNamesToProcess.add(name);
       }

--- a/common/changes/@microsoft/rush/stekycz-2551-includeDevDependencies_2021-03-13-23-00.json
+++ b/common/changes/@microsoft/rush/stekycz-2551-includeDevDependencies_2021-03-13-23-00.json
@@ -1,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "Fix \"rush deploy\" having \"includeDevDependencies\" turned on to deploy \"devDependencies\" for rush projects only",
+      "type": "none"
+    }
+  ],
+  "packageName": "@microsoft/rush",
+  "email": "martin.stekl@gmail.com"
+}

